pkg/action/build-uki.go

```diff
@@ -0,0 +1,538 @@
+package action
+
+import (
+	"compress/gzip"
+	"fmt"
+	"io"
+	"math"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"sort"
+	"strings"
+
+	"github.com/u-root/u-root/pkg/cpio"
+	"golang.org/x/exp/maps"
+
+	"github.com/kairos-io/enki/pkg/types"
+	"github.com/kairos-io/kairos-agent/v2/pkg/elemental"
+	v1 "github.com/kairos-io/kairos-agent/v2/pkg/types/v1"
+)
+
+const Cmdline = "console=ttyS0 console=tty1 net.ifnames=1 rd.immucore.oemlabel=COS_OEM rd.immucore.debug rd.immucore.oemtimeout=2 rd.immucore.uki selinux=0"
+
+type BuildUKIAction struct {
+	img           *v1.ImageSource
+	e             *elemental.Elemental
+	isoFile       string
+	keysDirectory string
+	logger        v1.Logger
+}
+
+func NewBuildUKIAction(cfg *types.BuildConfig, img *v1.ImageSource, result, keysDirectory string) *BuildUKIAction {
+	b := &BuildUKIAction{
+		logger:        cfg.Logger,
+		img:           img,
+		e:             elemental.NewElemental(&cfg.Config),
+		isoFile:       result,
+		keysDirectory: keysDirectory,
+	}
+	return b
+}
+
+func (b *BuildUKIAction) Run() error {
+	err := b.checkDeps()
+	if err != nil {
+		return err
+	}
+
+	b.logger.Info("extracting image to a temporary directory")
+	sourceDir, err := b.extractImage()
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(sourceDir)
+
+	b.logger.Info("creating additional directories")
+	if err := b.setupDirectoriesAndFiles(sourceDir); err != nil {
+		return err
+	}
+
+	b.logger.Info("creating an initramfs file")
+	if err := b.createInitramfs(sourceDir); err != nil {
+		return err
+	}
+
+	b.logger.Info("running ukify")
+	if err := b.ukify(sourceDir); err != nil {
+		return err
+	}
+
+	b.logger.Info("running sgsign")
+	if err := b.sbSign(sourceDir); err != nil {
+		return err
+	}
+
+	b.logger.Info("creating kairos and loader conf files")
+	if err := b.createConfFiles(sourceDir); err != nil {
+		return err
+	}
+
+	if err := b.createISO(sourceDir); err != nil {
+		return err
+	}
+
+	b.logger.Info(fmt.Sprintf("Done buidling the iso file: %s", b.isoFile))
+
+	return nil
+}
+
+func (b *BuildUKIAction) extractImage() (string, error) {
+	tmpDir, err := os.MkdirTemp("", "enki-build-uki-")
+	if err != nil {
+		return tmpDir, err
+	}
+
+	_, err = b.e.DumpSource(tmpDir, b.img)
+
+	return tmpDir, err
+}
+
+func (b *BuildUKIAction) checkDeps() error {
+	neededBinaries := []string{
+		"/usr/lib/systemd/ukify",
+		"sbsign",
+		"dd",
+		"mkfs.msdos",
+		"mmd",
+		"mcopy",
+		"xorriso",
+	}
+
+	neededFiles := []string{
+		"/usr/lib/systemd/boot/efi/linuxx64.efi.stub",
```

Comment:
this also needs change depending on the arch we are building for, this is for x86, arm64 has different stub+boot :(